segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3Storage.java

```diff
@@ -0,0 +1,384 @@
+    @Override
+    public CompletableFuture<Integer> read(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int
+            length, Duration timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncRead(handle, offset, buffer, bufferOffset, length));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> getStreamSegmentInfo(String streamSegmentName, Duration timeout) {
+        return supplyAsync(streamSegmentName, () -> syncGetStreamSegmentInfo(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(String streamSegmentName, Duration timeout) {
+        return supplyAsync(streamSegmentName, () -> syncExists(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openWrite(String streamSegmentName) {
+        return supplyAsync(streamSegmentName, () -> syncOpenWrite(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> create(String streamSegmentName, Duration timeout) {
+        return supplyAsync(streamSegmentName, () -> syncCreate(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle, long offset, InputStream data, int length, Duration
+            timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncWrite(handle, offset, data, length));
+    }
+
+    @Override
+    public CompletableFuture<Void> seal(SegmentHandle handle, Duration timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncSeal(handle));
+    }
+
+    @Override
+    public CompletableFuture<Void> concat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration
+            timeout) {
+        return supplyAsync(targetHandle.getSegmentName(),
+                () -> syncConcat(targetHandle, offset, sourceSegment));
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(SegmentHandle handle, Duration timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncDelete(handle));
+    }
+
+    //endregion
+
+    //region private sync implementation
+    private SegmentHandle syncOpenRead(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openRead", streamSegmentName);
+
+        StreamSegmentInformation info = syncGetStreamSegmentInfo(streamSegmentName);
+        ExtendedS3SegmentHandle retHandle = ExtendedS3SegmentHandle.getReadHandle(streamSegmentName);
+        LoggerHelpers.traceLeave(log, "openRead", traceId, streamSegmentName);
+        return retHandle;
+    }
+
+    private SegmentHandle syncOpenWrite(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+        StreamSegmentInformation info = syncGetStreamSegmentInfo(streamSegmentName);
+        ExtendedS3SegmentHandle retHandle;
+        if (info.isSealed()) {
+            retHandle = ExtendedS3SegmentHandle.getReadHandle(streamSegmentName);
```

Comment:
I have added a comment to the change to get a confirmation from @andreipaduroiu . it is important for me to have agreement over the contract. I personally find it awkward to return a read handle when I have asked for a write handle, but this semantics seems to be relatively harmless because the caller won't be able to do much damage with a read handle.